server/src-lib/Hasura/Tracing.hs

```diff
@@ -197,42 +202,64 @@ instance MonadTrace m => MonadTrace (ExceptT e m) where
   currentReporter = lift currentReporter
   attachMetadata = lift . attachMetadata
 
+-- | Create a Random 16 character hex string
+randomHex :: IO BS.ByteString
+randomHex = do
+  randNum <- liftIO (Rand.randomIO :: IO Word64)
+  let hexString = Hex.encode (BL.toStrict (BIN.encode  randNum))
+  return hexString
+
 -- | Inject the trace context as a set of HTTP headers.
 injectHttpContext :: TraceContext -> [HTTP.Header]
 injectHttpContext TraceContext{..} = 
-  [ ("X-Hasura-TraceId", fromString (show tcCurrentTrace))
-  , ("X-Hasura-SpanId", fromString (show tcCurrentSpan))
-  ]
+  case tcCurrentParent of
+    Nothing -> [ ("X-B3-TraceId", fromString (show tcCurrentTrace))
+               , ("X-B3-SpanId", fromString (show tcCurrentSpan))
+               ]
+    Just parentID ->  [ ("X-B3-TraceId", fromString (show tcCurrentTrace))
+                      , ("X-B3-SpanId", fromString (show tcCurrentSpan))
+                      , ("X-B3-ParentSpanId", fromString (show parentID))
+                      ]
+  
   
 -- | Extract the trace and parent span headers from a HTTP request
 -- and create a new 'TraceContext'. The new context will contain
 -- a fresh span ID, and the provided span ID will be assigned as
 -- the immediate parent span.
 extractHttpContext :: [HTTP.Header] -> IO (Maybe TraceContext)
 extractHttpContext hdrs = do
-  freshSpanId <- liftIO Rand.randomIO
+  freshSpanId <- liftIO randomHex
   pure $ TraceContext
-    <$> (HTTP.parseHeaderMaybe =<< lookup "X-Hasura-TraceId" hdrs)
+    <$> (HTTP.parseHeaderMaybe  =<< lookup "X-B3-TraceId" hdrs) 
     <*> pure freshSpanId
-    <*> pure (HTTP.parseHeaderMaybe =<< lookup "X-Hasura-SpanId" hdrs)
+    <*> pure (HTTP.parseHeaderMaybe =<< lookup "X-B3-SpanId" hdrs)
+    <*> pure Nothing
 
 -- | Inject the trace context as a JSON value, appropriate for 
 -- storing in (e.g.) an event trigger payload.
 injectEventContext :: TraceContext -> J.Value
 injectEventContext ctx =
-  J.object
-    [ "trace_id" J..= tcCurrentTrace ctx
-    , "span_id"  J..= tcCurrentSpan ctx
-    ]
+  case tcCurrentParent ctx of
```

Comment:
This part doesn't work as well because there is no instance of `J.ToJSON BS.ByteString`